File: dynamic-programming/knapsack/knapsack.kt
import java.lang.Integer.max
//input:
//  N: num of items
//  W: capacity of the bag
//  [V]: value of each item
//  [Wei]: weight of each item
//task:
//  maximize: ...
//  subject to: ...

fun main(args: Array<String>) {
    /* intput */
    val (N, W) = readLine()!!.split(" ").map{it.toInt()}
    var V: Array<Int> = Array(N) {0}
    var Wei: Array<Int> = Array(N) {0}
    for (index in V.indices) {
        val (a, b) = readLine()!!.split(" ").map{it.toInt()}
        V[index] = a
        Wei[index] = b
    }

    /* solve */
    var K: Array<Int> = Array(W+1) {0}
    for (w in 1..W) {
        for (i in 0..N-1) {
            if (w - Wei[i] >= 0) {
                K[w] = max(K[w], K[w - Wei[i]] + V[i])
            }
        }
    }
    
    /* output */
    println(K[W])
}

File: dynamic-programming/knapsack/knapsack.nim
import sequtils,strutils
import sugar

iterator range(n: int): int =
  for i in 0..<n:
    yield i

proc getLine[T](fun: string -> T): seq[T] =
  stdin.readLine.split.map fun

# input:
#   N: num of items
#   W: capacity of the bag
#   [V]: value of each item 
#   [Wei]: weight of each item
# task:
#   maximize: ...
#   subject to: ...

block main:
  # input
  var
     N,W: int
  (N,W) = getLine(parseInt)
  var V, Wei = newSeq[int](N)
  for i in range(N): 
    (V[i],Wei[i]) = getLine(parseInt)
  
  # solve
  var K = newSeq[int](W+1)
  for w in 1..W:
    for i in range(N):
      if w - Wei[i] >= 0:
        K[w] = max(K[w], K[w-Wei[i]] + V[i])
  
  # output
  K[W].echo
